### DISCUSSION TITLE
请问__open_real、__open_2、open分别会在哪些Android版本上被使用？

### DISCUSSION THREAD

**Discussion by zhbzhbzhbz** [bytedance/bhook#42] [Q&A]
如题，因为在sample中看到了这三个函数，而我使用frida-trace在Android9和Android12上，对FileOutputStream、FileWriter、openFileOutput、MediaStore进行探测时均只发现了对于open的调用~ 虽然我也看到了这个Issue中的回答，但还是不太确定https://github.com/bytedance/bhook/issues/16#issuecomment-976371103
谢谢您~！

**Comment by caikelun** [comment]
可以结合NDK里open的头文件，和AOSP源码看一下。具体我也记不得了。但是作为一个完整的针对“open”的PLT hook，目前需要hook这三个函数。

**Comment by zhbzhbzhbz** [comment]
最近发现，c语言下的`open(path, O_RDONLY);`用法会触发open2或者open_real，只使用open确实会拦截不到
